File: AuraConnect.Application/Services/GatewaySessionService.cs
using AuraConnect.Application.DTOs.Admin;
using AuraConnect.Application.DTOs.Gateway;
using AuraConnect.Application.Interfaces;
using AuraConnect.Core.Entities;
using AuraConnect.Core.Interfaces.Repositories;

namespace AuraConnect.Application.Services
{
    public class GatewaySessionService : IGatewaySessionService
    {
        private static readonly TimeSpan CorrelationWindow = TimeSpan.FromHours(1);

        private readonly IRadiusConfigRepository _radiusConfigRepository;
        private readonly IGatewaySessionEventRepository _gatewaySessionEventRepository;

        public GatewaySessionService(
            IRadiusConfigRepository radiusConfigRepository,
            IGatewaySessionEventRepository gatewaySessionEventRepository)
        {
            _radiusConfigRepository = radiusConfigRepository;
            _gatewaySessionEventRepository = gatewaySessionEventRepository;
        }

        public async Task<string> HandleEntryAsync(GatewayLoginRequest request, CancellationToken cancellationToken = default)
        {
            var host = ExtractHost(request.LinkLoginOnly) ?? ExtractHost(request.LinkStatus) ?? ExtractHost(request.LinkLogout);
            if (host == null)
                throw new InvalidOperationException("Could not determine the gateway host from the request");

            var config = await _radiusConfigRepository.GetByGatewayHostAsync(host, cancellationToken);
            if (config == null)
                throw new InvalidOperationException($"No site found for gateway host '{host}'");

            var site = config.Site;
            if (string.IsNullOrWhiteSpace(site.Domain))
                throw new InvalidOperationException("Site has no portal domain configured");

            var redirectUrl = BuildPortalUrl(site, "splash");

            var gatewaySessionEvent = new GatewaySessionEvent(
                site.Id, host, redirectUrl,
                request.Mac, request.NasId, request.LinkLoginOnly, request.LinkStatus, request.LinkLogout);

            await _gatewaySessionEventRepository.AddAsync(gatewaySessionEvent, cancellationToken);
            await _gatewaySessionEventRepository.SaveChangesAsync(cancellationToken);

            return $"{redirectUrl}?session={gatewaySessionEvent.Id}";
        }

        public async Task<string?> RecordLoginResultAsync(GatewayLoginResultRequest request, CancellationToken cancellationToken = default)
        {
            if (string.IsNullOrWhiteSpace(request.Mac))
                return null;

            var gatewaySessionEvent = await _gatewaySessionEventRepository.FindMostRecentByMacAsync(request.Mac, CorrelationWindow, cancellationToken);
            if (gatewaySessionEvent == null)
                return null;

            var outcome = string.Equals(request.Result, "success", StringComparison.OrdinalIgnoreCase)
                ? GatewayLoginOutcome.Success
                : GatewayLoginOutcome.Failed;

            gatewaySessionEvent.RecordLoginOutcome(outcome, request.Error, request.ErrorOriginal);

            await _gatewaySessionEventRepository.UpdateAsync(gatewaySessionEvent, cancellationToken);
            await _gatewaySessionEventRepository.SaveChangesAsync(cancellationToken);

            var statusUrl = BuildPortalUrl(gatewaySessionEvent.Site, "splash") + $"?status={(outcome == GatewayLoginOutcome.Success ? "success" : "failed")}";
            if (outcome == GatewayLoginOutcome.Failed && !string.IsNullOrWhiteSpace(request.Error))
                statusUrl += $"&reason={Uri.EscapeDataString(request.Error)}";

            return statusUrl;
        }

        public async Task<PagedResult<GatewaySessionEventResponse>> GetSessionsAsync(
            int page, int pageSize, string? tenantId, string? siteId, string? mac, GatewayLoginOutcome? outcome,
            DateTime? from, DateTime? to, CancellationToken cancellationToken = default)
        {
            var (items, total) = await _gatewaySessionEventRepository.GetPagedAsync(
                page, pageSize, tenantId, siteId, mac, outcome, from, to, cancellationToken);

            return new PagedResult<GatewaySessionEventResponse>
            {
                Items = items.Select(MapToResponse),
                Page = page,
                PageSize = pageSize,
                TotalCount = total
            };
        }

        private static string? ExtractHost(string? url) =>
            Uri.TryCreate(url, UriKind.Absolute, out var uri) ? uri.Host : null;

        private static string BuildPortalUrl(Site site, string path) =>
            $"https://{site.Domain}/{Uri.EscapeDataString(site.Ssid)}/{path}";

        private static GatewaySessionEventResponse MapToResponse(GatewaySessionEvent g) => new()
        {
            Id = g.Id,
            SiteId = g.SiteId,
            Mac = g.Mac,
            NasId = g.NasId,
            ResolvedHost = g.ResolvedHost,
            RedirectUrl = g.RedirectUrl,
            LoginOutcome = g.LoginOutcome.ToString(),
            LoginError = g.LoginError,
            LoginErrorOriginal = g.LoginErrorOriginal,
            LoginCompletedAt = g.LoginCompletedAt,
            CreatedAt = g.CreatedAt
        };
    }
}

File: clients/current/admin/src/lib/services/gateway-session.service.ts
import { gatewaySessionsApi } from '@/lib/infrastructure/api/gateway-sessions.api'
import type { GatewaySessionListParams as ApiParams } from '@/lib/infrastructure/api/gateway-sessions.api'
import type { components } from '@/lib/infrastructure/api/schema'
import type { GatewaySessionEvent, PagedGatewaySessions, GatewaySessionListParams } from '@/lib/types/gateway-session.types'

type ApiEvent = components['schemas']['GatewaySessionEventResponse']

function decodeMac(mac: string | null | undefined): string | null {
  if (!mac) return null
  try {
    return decodeURIComponent(mac)
  } catch {
    return mac
  }
}

function toGatewaySessionEvent(api: ApiEvent): GatewaySessionEvent {
  return {
    id: api.id ?? '',
    siteId: api.siteId ?? '',
    mac: decodeMac(api.mac),
    nasId: api.nasId ?? null,
    resolvedHost: api.resolvedHost ?? null,
    redirectUrl: api.redirectUrl ?? null,
    loginOutcome: api.loginOutcome ?? 'Pending',
    loginError: api.loginError ?? null,
    loginErrorOriginal: api.loginErrorOriginal ?? null,
    loginCompletedAt: api.loginCompletedAt ?? null,
    createdAt: api.createdAt ?? new Date().toISOString(),
  }
}

export const gatewaySessionService = {
  async getAll(params: GatewaySessionListParams = {}): Promise<PagedGatewaySessions> {
    const apiParams: ApiParams = {
      page: params.page,
      pageSize: params.pageSize,
      tenantId: params.tenantId,
      siteId: params.siteId,
      mac: params.mac,
      outcome: params.outcome,
      from: params.from,
      to: params.to,
    }
    const paged = await gatewaySessionsApi.getAll(apiParams)
    return {
      items: (paged.items ?? []).map(toGatewaySessionEvent),
      page: Number(paged.page ?? 1),
      pageSize: Number(paged.pageSize ?? 25),
      totalCount: Number(paged.totalCount ?? 0),
      totalPages: Number(paged.totalPages ?? 1),
      hasNextPage: paged.hasNextPage ?? false,
      hasPreviousPage: paged.hasPreviousPage ?? false,
    }
  },
}
